### PR TITLE
Switch resourcequota controller to shared informers

### PR DESCRIPTION
**Pull request by ncdc**
Originally part of #40097 

I have had some issues with this change in the past, when I updated `pkg/quota` to use the new informers while `pkg/controller/resourcequota` remained on the old informers. In this PR, both are switched to using the new informers. The issues in the past were lots of flakey test failures in the ResourceQuota e2es, where it would randomly fail to see deletions and handle replenishment. I am hoping that now that everything here is consistently using the new informers, there won't be any more of these flakes, but it's something to keep an eye out for.

I also think `pkg/controller/resourcequota` could be cleaned up. I don't think there's really any need for `replenishment_controller.go` any more since it's no longer running individual controllers per kind to replenish. It instead just uses the shared informer and adds event handlers to it. But maybe we do that in a follow up.

cc @derekwaynecarr @smarterclayton @wojtek-t @deads2k @sttts @liggitt @timothysc @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

**Comment by ncdc** [comment]
I've got a fix for the integration test locally, but I do see more quota failures in the e2e. I haven't been able to reproduce those locally so far but I'll keep trying. @derekwaynecarr if you have any spare time I could use another set of eyes.

**Comment by ncdc** [comment]
Ok, the e2e failures look like an issue with the bootstrap policy. Will fix asap.

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 1]
```diff
@@ -262,17 +270,16 @@ func (rq *ResourceQuotaController) syncResourceQuotaFromKey(key string) (err err
...
 
 // syncResourceQuota runs a complete sync of resource quota status across all known kinds
-func (rq *ResourceQuotaController) syncResourceQuota(v1ResourceQuota v1.ResourceQuota) (err error) {
+func (rq *ResourceQuotaController) syncResourceQuota(v1ResourceQuota *v1.ResourceQuota) (err error) {
 	// quota is dirty if any part of spec hard limits differs from the status hard limits
 	dirty := !apiequality.Semantic.DeepEqual(v1ResourceQuota.Spec.Hard, v1ResourceQuota.Status.Hard)
 
 	resourceQuota := api.ResourceQuota{}
-	if err := v1.Convert_v1_ResourceQuota_To_api_ResourceQuota(&v1ResourceQuota, &resourceQuota, nil); err != nil {
+	if err := v1.Convert_v1_ResourceQuota_To_api_ResourceQuota(v1ResourceQuota, &resourceQuota, nil); err != nil {
```
Comment (deads2k) [inline pkg/controller/resourcequota/resource_quota_controller.go]: This doesn't do a deep copy.  Bet you didn't see *that* coming.

Have you reasoned about safety the rest of the way down?

**Comment by deads2k** [comment]
@derekwaynecarr you definitely need to look at this one.

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 1 above)
Comment (ncdc) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Argh I assumed it did, so I chose not to copy. I can do so.

**Comment by ncdc** [comment]
I have to see what (if anything) in this PR is causing the kubelet not to start.

File: pkg/controller/resourcequota/replenishment_controller.go [hunk 2]
```diff
@@ -273,18 +196,3 @@ func IsUnhandledGroupKindError(err error) bool {
 	_, ok := err.(unhandledKindErr)
 	return ok
 }
-
-// UnionReplenishmentControllerFactory iterates through its constituent factories ignoring, UnhandledGroupKindErrors
-// returning the first success or failure it hits.  If there are no hits either way, it return an UnhandledGroupKind error
-type UnionReplenishmentControllerFactory []ReplenishmentControllerFactory
```
Comment (derekwaynecarr) [inline pkg/controller/resourcequota/replenishment_controller.go]: why was this removed?

Comment (derekwaynecarr) [inline pkg/controller/resourcequota/replenishment_controller.go]: if i recall, we wanted to use this in a future pr when we added other groups.

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 3]
```diff
@@ -228,18 +231,19 @@ func (rq *ResourceQuotaController) worker(queue workqueue.RateLimitingInterface)
...
 	defer utilruntime.HandleCrash()
-	go rq.rqController.Run(stopCh)
-	// the controllers that replenish other resources to respond rapidly to state changes
-	for _, replenishmentController := range rq.replenishmentControllers {
-		go replenishmentController.Run(stopCh)
+
+	glog.Infof("Starting resource quota controller")
+
+	if !cache.WaitForCacheSync(stopCh, rq.informerSyncedFuncs...) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
```
Comment (derekwaynecarr) [inline pkg/controller/resourcequota/resource_quota_controller.go]: what happens if we timeout?  it looks like we fail silently...

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 4]
```diff
@@ -279,13 +279,25 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule("get").Groups(legacyGroup).Resources("endpoints", "namespaces", "serviceaccounts").RuleOrDie(),
 				rbac.NewRule("update").Groups(legacyGroup).Resources("endpoints", "serviceaccounts").RuleOrDie(),
 
-				rbac.NewRule("list", "watch").Groups("*").Resources("namespaces", "nodes", "persistentvolumeclaims",
-					"persistentvolumes", "pods", "secrets", "services", "serviceaccounts", "replicationcontrollers").RuleOrDie(),
+				rbac.NewRule("list", "watch").Groups(legacyGroup).Resources(
```
Comment (derekwaynecarr) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: i am confused why this pr needs to touch here.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
same comment as david, just a few questions throughout.

File: pkg/controller/resourcequota/replenishment_controller.go (hunk 2 above)
Comment (ncdc) [inline pkg/controller/resourcequota/replenishment_controller.go]: Since I was deleting the concept of a replenishment controller factory, I got rid of this too. I can see if there's some type I can add back that will achieve a similar effect.

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 3 above)
Comment (ncdc) [inline pkg/controller/resourcequota/resource_quota_controller.go]: This logs by default. This is consistent with all the other controllers.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go (hunk 4 above)
Comment (ncdc) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: Because the kube controller manager's identity is the client that is used to run the shared informers, and it needs to be able to list/watch everything used by all the controllers.

Comment (derekwaynecarr) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: ok, that makes sense now.

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 1 above)
Comment (ncdc) [inline pkg/controller/resourcequota/resource_quota_controller.go]: I think we are ok without a deep copy. I believe that all of the quota manipulation (`quota.Add`, `quota.Mask`) works on copies of the data, and when we create the `ResourceQuota` object we use for updating the status, we assign the original quota's annotations and labels to it, but I don't see any issue doing that.

**Comment by ncdc** [comment]
@liggitt this is the PR that inexplicably keeps the kubelet from starting up, saying it can't find any authentication methods configured.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml [hunk 5]
```diff
@@ -444,14 +444,16 @@ items:
     verbs:
     - update
   - apiGroups:
-    - '*'
+    - ""
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml]: are you sure you're rebased on master? latest master doesn't have this.

Comment (ncdc) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml]: Locally I have rebased. I haven't pushed yet. I did figure out my issue though.

File: pkg/controller/resourcequota/replenishment_controller.go (hunk 2 above)
Comment (ncdc) [inline pkg/controller/resourcequota/replenishment_controller.go]: Added back in latest iteration

**Comment by ncdc** [comment]
Rebased, updated, brought back the union replenishment controller factory. PTAL.

**Comment by ncdc** [comment]
@k8s-bot gci gke e2e test this

One of the heapster pods wouldn't start up. Not sure why without master logs. Let's see if it was some random flake...

**Comment by ncdc** [comment]
Woohoo all the tests passed this time 😄

File: pkg/controller/resourcequota/resource_quota_controller.go (hunk 1 above)
Comment (derekwaynecarr) [inline pkg/controller/resourcequota/resource_quota_controller.go]: sgtm

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 6]
```diff
@@ -228,18 +236,24 @@ func (rq *ResourceQuotaController) worker(queue workqueue.RateLimitingInterface)
...
-	go rq.rqController.Run(stopCh)
+
+	glog.Infof("Starting resource quota controller")
+
 	// the controllers that replenish other resources to respond rapidly to state changes
 	for _, replenishmentController := range rq.replenishmentControllers {
 		go replenishmentController.Run(stopCh)
 	}
+
+	if !cache.WaitForCacheSync(stopCh, rq.informerSyncedFuncs...) {
```
Comment (derekwaynecarr) [inline pkg/controller/resourcequota/resource_quota_controller.go]: can you add a comment that explains why its OK for this to just return?

Comment (derekwaynecarr) [inline pkg/controller/resourcequota/resource_quota_controller.go]: if it will be a common pattern for every other controller, we can do it in a follow-on and possible put it behind a shared library call.  my understanding is that the wait is infinite, unless it gets something on the stopCh which is a shutdown of controller anyway.

Comment (ncdc) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Yes, it's common for all controllers, and the wait is infinite until stopCh is closed (process terminating). Let's do it in a follow-up.

**Review by derekwaynecarr — APPROVED** [review]
